src/Open3D/Core/Kernel/ReductionCUDA.cu

```diff
@@ -0,0 +1,1101 @@
+            } else {
+                if (accumulate_) {
+                    value = ops_.Combine(*acc, value);
+                }
+                if (final_output_) {
+                    SetResultsToOutput(value, base_offsets[0]);
+                } else {
+                    *acc = value;
+                }
+            }
+        }
+    }
+
+    OPEN3D_DEVICE arg_t ThreadReduce(const scalar_t* data) const {
+        index_t idx = config_.InputIdx();
+        // Multiple accumulators to remove dependency between unrolled loops.
+        arg_t value_list[vt0];
+#pragma unroll
+        for (int i = 0; i < vt0; i++) {
+            value_list[i] = identity_;
+        }
+        index_t end = config_.num_inputs_per_output_;
+        index_t stride = config_.step_input_;
+        index_t element_stride = input_calc_.strides_[0][0] / sizeof(scalar_t);
+
+        // Reducing layers of function calls so compiler could do proper loop
+        // unroll that exposes instruction level parallelism.
+        while (idx < config_.num_inputs_per_output_) {
+            // load input
+            SmallArray<scalar_t, vt0> values;
+            if (input_calc_.dims_ == 1) {
+                StridedIterate<vt0>(
+                        [&](index_t i, index_t idx) {
+                            values[i] = data[idx * element_stride];
+                        },
+                        idx, end, stride);
+            } else {
+                StridedIterate<vt0>(
+                        [&](index_t i, index_t idx) {
+                            values[i] = data[input_calc_.get(idx)[0] /
+                                             sizeof(scalar_t)];
+                        },
+                        idx, end, stride);
+            }
+            // compute
+            StridedIterate<vt0, index_t>(
+                    [&](index_t i, index_t idx) {
+                        value_list[i] =
+                                ops_.Reduce(value_list[i], values[i], idx);
+                    },
+                    idx, config_.num_inputs_per_output_, config_.step_input_);
+            // step offset
+            idx += config_.step_input_ * vt0;
+        }
+#pragma unroll
+        for (int i = 1; i < vt0; i++) {
+            value_list[0] = ops_.Combine(value_list[0], value_list[i]);
+        }
+        return value_list[0];
+    }
+
+    OPEN3D_DEVICE arg_t BlockXReduce(arg_t value, char* shared_memory) const {
+        int dim_x = blockDim.x;
+        arg_t* shared = (arg_t*)shared_memory;
+        if (dim_x > warpSize) {
+            int address_base = threadIdx.x + threadIdx.y * blockDim.x;
+            shared[address_base] = value;
+            for (int offset = dim_x / 2; offset >= warpSize; offset >>= 1) {
+                __syncthreads();
+                if (threadIdx.x < offset && threadIdx.x + offset < blockDim.x) {
+                    arg_t other = shared[address_base + offset];
+                    value = ops_.Combine(value, other);
```

Comment:
```cpp
int dim_x = blockDim.x;
arg_t* shared = (arg_t*)shared_memory;
if (dim_x > warpSize) {
    int address_base = threadIdx.x + threadIdx.y * blockDim.x;
    shared[address_base] = value;
    for (int offset = dim_x / 2; offset >= warpSize; offset >>= 1) {
        __syncthreads();
        if (threadIdx.x < offset && threadIdx.x + offset < blockDim.x) {
            arg_t other = shared[address_base + offset];
            value = ops_.Combine(value, other);
            shared[address_base] = value;
        }
    }
    dim_x = warpSize;
}
```

This case shouldn't happen since `shared[address_base + offset]` won't overlap with `shared[address_base]` for all possible threads in the block.

- Case 1: `thread_a. threadIdx.y` != `thread_b. threadIdx.y`.
  In this case:
  - `thread_a.address_base` and `thread_b.address_base` are at least `blockDim.x` elements apart.blockDim.x
  - `offset < blockDim.x`
  - Therefore, `shared[thread_a.address_base + offset]` won't overlap with `shared[thread_b.address_base]`.
- Case 2: `thread_a. threadIdx.y` == `thread_b. threadIdx.y`.
  Inside the condition `if (threadIdx.x < offset ...)`:
  ````
    a.address_base + offset
  > a.address_base + b.threadIdx.x
  > a.address_base + b.threadIdx.x - a.threadIdx.x
  = a.address_base - a.threadIdx.x + b.threadIdx.x
  = b.address_base
  ````
  Therefore `shared[address_base + offset]` is always after `shared[address_base]` for all possible threads.

Reply:
oh yes, threadIdx.x<offset is in the if, no __syncthreads() is needed